### PR TITLE
SIR in IEWPF

### PR DESCRIPTION
**Pull request by havahol**
This PR is based on top of PR #177, and I therefore close #177.

From PR #177:
- Improving plots for paper.
- Creating figure for forecast error statistics.
- Fixing bug related to the eta-compensation.

Additional notebooks in this PR:
- Notebook with figures for computational performance
- Investigating optimal block size configuration for the model with mod…  …
- Investigating the SIR particle filter and its collapse

Other additions to this PR:
- Improved normalization of weights required to make SIR work for the cases investigated recently
- Some performance optimization for bicubic interpolation
- Minor improvements related to figures and notebooks.

### CODE REVIEW

File: gpu_ocean/demos/DAPaper/scripts/run_model.py [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""
+
```
Comment (setmar) [inline gpu_ocean/demos/DAPaper/scripts/run_model.py]: Add a comment on the purpose of this script.

Comment (havahol) [reply gpu_ocean/demos/DAPaper/scripts/run_model.py]: Already at lines 9-11

File: gpu_ocean/SWESimulators/gpu_kernels/ocean_noise.cu [hunk 2]
```diff
@@ -581,43 +581,48 @@ __global__ void bicubicInterpolation(
...
+                f_matrix.m_row3.y = (coarse[loc_j+1][loc_i+2] - coarse[loc_j+1][loc_i  ])/2.0f; // fx11
+
+                // Derivatives in y-direction in the corner points
+                f_matrix.m_row0.z = (coarse[loc_j+1][loc_i  ] - coarse[loc_j-1][loc_i  ])/2.0f; // fy00
+                f_matrix.m_row0.w = (coarse[loc_j+2][loc_i  ] - coarse[loc_j  ][loc_i  ])/2.0f; // fy01
+                f_matrix.m_row1.z = (coarse[loc_j+1][loc_i+1] - coarse[loc_j-1][loc_i+1])/2.0f; // fy10
+                f_matrix.m_row1.w = (coarse[loc_j+2][loc_i+1] - coarse[loc_j  ][loc_i+1])/2.0f; // fy11
+
+                // Derivatives in y-direction required for obtaining the cross derivatives
+                const float fy_10 = (coarse[loc_j+1][loc_i-1] - coarse[loc_j-1][loc_i-1])/2.0f; // fy_10
```
[the remark this answers is not in the document]
Comment (havahol) [reply gpu_ocean/SWESimulators/gpu_kernels/ocean_noise.cu]: Adding following explanation
```
                // We use the following notation:
                //     function values (f) or derivatives (fx, fy), or cross derivatives (fxy)
                //     Evaluated in the unit square at given positions
                //     e.g., f01   = function value at (0, 1)
                //           fy_10 = value of derivative wrt y at (-1,0)
                //           fy21  = value of derivative wrt y at ( 2,1)
```

File: gpu_ocean/SWESimulators/Observation.py [hunk 3]
```diff
@@ -149,9 +153,17 @@ def add_observation_from_sim(self, sim):
...
-                buoy_observations[i, 0] = hu[self.buoy_indices[i,1], self.buoy_indices[i,0]]
-                buoy_observations[i, 1] = hv[self.buoy_indices[i,1], self.buoy_indices[i,0]]
+                # buoy index
+                x = self.buoy_indices[i, 0]
+                y = self.buoy_indices[i, 1]
+                
+                # Observe current, and we know the depth, but not eta
+                eta_ignorance_factor = H[y, x] / (H[y, x] + eta[y, x])
+                buoy_observations[i, 0] = hu[y, x] * eta_ignorance_factor
+                buoy_observations[i, 1] = hv[y, x] * eta_ignorance_factor
```
[the remark this answers is not in the document]
Comment (havahol) [reply gpu_ocean/SWESimulators/Observation.py]: Correct

**Comment by havahol** [comment]
All comments followed up - self-merging

**Pull request merged**